examples/structured_data/feature_preprocessing_layer.py

```diff
@@ -0,0 +1,224 @@
+        # set preprocessing layers
+        self.normalizer = Normalization()
+        self.categoric_string_lookup = StringLookup()
+        self.integer_lookup = IntegerLookup(output_mode="binary")
+        self.discretizer = Discretization(num_bins=10)
+        self.embedding_string_lookup = StringLookup()
+
+    def adapt(self, data={}):
+        """Update layer states"""
+        self.normalizer.adapt(data["numerical_features"])
+
+        self.categoric_string_lookup.adapt(data["categorical_features"])
+
+        # apply integer lookup on string lookup layer output
+        self.integer_lookup.adapt(
+            self.categoric_string_lookup(data["categorical_features"])
+        )
+
+        self.discretizer.adapt(data["discretize_features"])
+
+        self.embedding_string_lookup.adapt(data["embedding_features"])
+
+        # create sequential model for string and integer lookup layers
+        self.categoric_encoding = tf.keras.models.Sequential(
+            [self.categoric_string_lookup, self.integer_lookup]
+        )
+
+        vocabulary_size = self.embedding_string_lookup.vocabulary_size()
+        embedding_size = int(tf.math.sqrt(vocabulary_size))
+        # create sequential model for embedding feature
+        self.embedding = tf.keras.models.Sequential(
+            [
+                self.embedding_string_lookup,
+                Embedding(input_dim=vocabulary_size, output_dim=embedding_size),
+                Reshape((-1,)),
+            ]
+        )
+
+    def call(self, inputs={}):
+        """Apply adapted layers"""
+        numerical_features = self.normalizer(inputs["numerical_features"])
+
+        categorical_features = self.categoric_encoding(inputs["categorical_features"])
+
+        discretize_features = tf.cast(
+            self.discretizer(inputs["discretize_features"]), dtype=tf.float32
+        )
+
+        embedding_features = self.embedding(inputs["embedding_features"])
+
+        return tf.keras.layers.concatenate(
+            [
+                numerical_features,
+                categorical_features,
+                discretize_features,
+                embedding_features,
+            ]
+        )
+
+
+"""
+## Compile and fit model
+
+We now have a feature preprocessing layer. Lets put everything together and fit the model:
+"""
+
+inputs = {
+    "numerical_features": tf.keras.Input(
+        shape=(len(NUMERIC_FEATURES),), dtype=tf.float32, name="numeric_layer"
+    ),
+    "categorical_features": tf.keras.Input(
+        shape=(len(CATEGORICAL_FEATURES),), dtype=tf.string, name="categoric_layer"
+    ),
+    "discretize_features": tf.keras.Input(
+        shape=(len(DISCRETIZE_FEATURES),), dtype=tf.float32, name="discretize_layer"
+    ),
+    "embedding_features": tf.keras.Input(
+        shape=(len(EMBEDDING_FEATURES),), dtype=tf.string, name="embedding_layer"
+    ),
+}
+
+# update preprocessing layer states
+preprocessing_layer = FeaturePreprocessor()
```

Comment:
Let's use the common pattern where during training we `map()` the preprocessor into the tf.data pipeline, and at test time, we create a model that includes the preprocessor.

e.g.
```
input_dict = ...
preprocessed_features = preprocessor(input_dict)
x = layers.Dense(64, activation="relu")(preprocessed_features)
outputs = ...

training_model = keras.Model(inputs=preprocessed_features, outputs=outputs)
inference_model = keras.Model(inputs=input_dict, outputs=outputs)
```

Also let's explain that we do this because we want to do asynchronous parallel preprocessing at training time.

Reply:
I have implemented this pattern on the changes.

Reply:
Hi @fchollet . This is quite an old PR and I have updated it now with all the suggestions and changes requested. Apologies that I only updated it now. Please have a look and inform me if we can still work on this.